### ISSUE TITLE
What does create-slave do?

### ISSUE THREAD

**Issue by FrankC01** [vygr/ChrysaLisp#73]
I've started reviewing and want to start with adding a few command to the TUI. I came upon the function but not sure what it is doing...

Thanks

**Comment by vygr** [comment]
It creates an instance of the class/slave. This class waits for the pipeline of process started by the Terminal to start and be wired up. Waits for the args to be received and sets up stdin, stdout and stderr message streams. 

Take a look at the Terminal apps pipe.inc file to see how the command line is launched and the class/slave init function to see it cooperating with the Terminal. 

It think maybe it’s badly named and I should call it something else maybe ?

Regards 

Chris

**Comment by FrankC01** [comment]
OK - first thought was stdin, etc. would be automagically available so it was confusing. I have not yet grasped the overall yet and I have a steep curve with lisp in general (previously did lots with Clojure).  Perhaps 'context' may be a name to use as it describes a terminal context ... just a thought.

Also, what is the difference between `defun` and `defun-bind`?

**Comment by vygr** [comment]
Defun is the usual macro to define a named lambda. 

Defun-bind is pretty much the same but it also runs bind-fun on the lambda body which pre-binds all symbols that evaluate to named functions. This save looking up those bindings at runtime. 

All these standard macros and built ins are defined in the class/lisp/boot.inc file. Take a browse through that to see many base function implantations. 

Chris

**Comment by vygr** [comment]
There is no stdin, stdout, stderr as such. They are just a model that the Terminal app and cmd/ apps use.

Under the hood there is only messages and processes. Any concept like std streams is purely a useful model for this type of pipelines of processes programming style. But it’s a very useful  one for terminal style tools etc.

**Comment by FrankC01** [comment]
Got it. I will do some deeper reading.

**Comment by vygr** [comment]
Any questions just ask.

I’ll be asleep for a while now :) catch up after breakfast! About 8 hours time.

**Comment by vygr** [comment]
How about renaming class/slave to class/stdio as that’s what’s its doing setting up your process to behave as a stdio process. 

And the Lisp binding renamed to create-stdio ?

Chris

**Comment by FrankC01** [comment]
Works for me and wears it's semantics on it's sleeve

**Comment by vygr** [comment]
Pushed that rename and another that was on my list to change.

**Comment by FrankC01** [comment]
Looks good... I'm thinking of a more sophisticated argument parser would be a good first toe in the water...

**Comment by vygr** [comment]
Absolutely!!! The one I did was just to get something going. 

But before a lot of cmd apps are done a more sophisticated arg parser would benefit everyone downstream.

**Comment by FrankC01** [comment]
OK, I'll keep chipping away. I did something in a language I wrote and also think Python's argparse has an interesting take... I'll propose something more concrete in the upcoming days.

**Comment by FrankC01** [comment]
So my initial thoughts, for a first pass, is a structure that captures a 'tree' of commands. Let me know if you can't see the image. Now I don't see a `defclass` option so I can probably define a hash-map (`hmap`?). The top level object contains general application information and a collection of `cmds` each, in turn, can hold subcommands.

![chrysargs-3](https://user-images.githubusercontent.com/5435165/81792907-afb4df00-94d6-11ea-8450-2acf167b8cba.png)

**Comment by vygr** [comment]
The (env) function let’s you create isolated hmaps. Like cmd/oops.lisp does. Or you can have them have references back to the parent environment if you need to. 

Or you may create a simple flat map with a pair of lists one holding a symbol as key and the other the values. (Find) has and optimised fast scan for a symbol in a list of symbols so you can build a very fast flat map this way. 

Clearly you could also have multiple buckets too. I do combos of these simple symbol maps in various places.

**Comment by vygr** [comment]
The environment is just a chain of these symbol maps just with a few extra assembler level helpers. But you may get more control and flexibility with one you roll yourself ?

Are you going to have any issue with cycles in the sub commands? Can they reference up or will it really be a tree ?

There is no GC so this might be a Gotcha to watch out for.

**Comment by FrankC01** [comment]
I can remove the cycle issue but, for now, going with a flat model where the `chryscmds` don't have children.

**Comment by FrankC01** [comment]
Quick question: The `(structure name...)` looks inviting and I've created one using `ptr` type members but it isn't clear how to create an instance of a structure?

For example, I'm thinking each 'chrycmd' is an instance of the structure. 

What am I missing here?

**Comment by vygr** [comment]
Structure just creates symbol equates. Used for creating event numbers or tupple element names or message structures. 

It’s not something you can instance as such. More a helper in naming things. 

A list used as a tupple, elements named with structure and accessed with the (tupple-get) and (tupple-set) macros to pre evaluate the element offsets is used quite often as an instance.

**Comment by vygr** [comment]
Structure at the VP level is a different thing. This is used to define the format of VP level instance data. VP classes in the class library and gui etc use ref counted instances of structures as objects.

**Comment by vygr** [comment]
The (get-field) and (set-field) functions give you a way to access VP instance data fields from Lisp code. This is used often in the Lisp bindings to VP level classes. 

In your case for the option handling I think your probably best served by a Tupple structure (if you have a fixed set of elements) or symbol hash idea (if you need varying number of elements).

**Comment by FrankC01** [comment]
OK, and keywords `:foo` are not supported at this time correct?

**Comment by vygr** [comment]
Other things to consider is how do you wish evaluation to work in any structure you use ? Are you just getting and setting fields or are you wanting to (eval) symbols and have the Lisp evaluation just work (tm) ?

If you want eval to just work you need to build your tree or chain with (env) in a similar way to the oops.lisp code and the GUI components. Then (eval form e) can be used and the Lisp machinery will do the heavy lifting for you.

**Comment by FrankC01** [comment]
Likely a combination, what I'm striving for at the moment while  azstill getting over the curve of CLisp (yours) is to have a rudimentary working set and I will post a Gist for you to take a look.

**Comment by vygr** [comment]
Keywords ? :foo is the same as ‘foo. 

Just a symbol. 

‘(x y z)

Would be a list of symbols x y z. 

Finding a symbol in a list of symbols. 

(find ‘a ‘(x y a)) => 2

**Comment by vygr** [comment]
Symbols are interned unlike strings. So ‘a would always be the exact same object as any other ‘a. 

That’s not true of “a”.

**Comment by FrankC01** [comment]
Yes, keywords are just a personal visual/readability thing... I'm good with straight symbols.

**Comment by vygr** [comment]
Something that might be useful and maybe more along the lines of :foo (I might be wrong here) is a symbol that Evals to itself. 

I do this in the VP assembler for VP register symbols and this might be somthing useful as they pass through evaluation levels without changing. 

(defq ‘a ‘a)

That may fit more with the :a use ?

**Comment by vygr** [comment]
If that is of use I could add : to the (read) function as a shortcut for doing this ?

**Comment by FrankC01** [comment]
Yes `(defq 'a 'a)` will do.

Fundamentally:
Stealing from oops.lisp construct a base class (app name, version, help text, etc.), argument or switch class ("-f", "--file") and a command class ("cmdA", "cmdB") that can be contained by the base class. 'args' can be type verified and 'cmds' can contain functions to call after parse validations.

> foo read -f cmds/sample.inc -r 20 

Where `read` is a command that takes an argument `-f` for file reference and `-r` to read the first 20 lines. `f` can be validated as a valid file and `-r` validate value is an integer. Internally, the `read` may have a function to pass control to after argument validation, passing a class exposes the values of the arguments.

Something like that

**Comment by FrankC01** [comment]
```
(class _base)
(method _base 'init)
(method _base 'deinit)
(method _base 'set_help)
(method _base 'get_help)

(class _arg _base)
(method _arg 'set_switch)
(method _arg 'get_switch)
(method _arg 'set_type)
(method _arg 'get_type)
(method _arg 'validate)

(class _cmd _base)
(method _cmd 'set_cmd)
(method _cmd 'get_cmd)
(method _cmd 'add_arg)
(method _cmd 'args)
(method _cmd 'set_fn)
(method _cmd 'get_fn)
(method _cmd 'validate) ; this is walked during ChrysAP validate

(class ChrysAP _base)
(method ChrysAP 'set_appname)
(method ChrysAP 'get_appname)
(method ChrysAP 'set_version)
(method ChrysAP 'get_version)
(method ChrysAP 'add_cmd)
(method ChrysAP 'cmds)
(method ChrysAP 'add_arg)
(method ChrysAP 'args)
(method ChrysAP 'validate) ; this is what is called with cmd line input

```

**Comment by vygr** [comment]
Looks good so far :)

**Comment by FrankC01** [comment]
Just getting fundamentals in: [VIEW HERE](https://gist.github.com/FrankC01/fb30f3cc4baa5e8eb4a6a43c41d57e4d)

**Comment by vygr** [comment]
:)

**Comment by vygr** [comment]
So here:

(method ChArg 'deinit (lambda (o)
	(set o 'argtext nil)
	(set o 'atype nil)
	(scall ChArg 'deinit o)))

Can do:

(set o ’argtext nil ‘atype nil)

Minor stuff but worth knowing.

**Comment by FrankC01** [comment]
Yes, I did see that however; for my own aging eyes for readability I'm doing something like:

```
(set o 'argtext nil
       o 'atype nil)

```
Which brings up another point, are their constants for `types` ? I know there are `(lst? ...)` and others but I would like the user to provide a constant if there are....

`(type-of x) returns a numeric value`

**Comment by vygr** [comment]
Look at the very top of class/lisp/boot.inc where those type testing macros are defined you will see the constants that (type) returns.

**Comment by vygr** [comment]
Around line 117

(defq t 't nil 'nil
	lisp_type_lst (<< 1 0)
	lisp_type_int (<< 1 1)
	lisp_type_str (<< 1 2)
	lisp_type_sym (<< 1 3)
	lisp_type_fnc (<< 1 4)
 	defmacro '(macro (n a &rest _) `(defq ,n (macro ,a ~_))))

**Comment by vygr** [comment]
Now I might need to add some more since I added these as several more classes have gone in since then. 

I find and try to avoid code that tests for the type of something, but sometimes its just too nice to have polymorphic activity without relying on the VP level Lisp bindings doing it for you. 

And I can sure see that you need to be able to do type checking in an argument parser !

**Comment by vygr** [comment]
Class/lisp/lisp_type.lisp is the file to play with if you want to add some more to the few I already have if you need to experiment.

**Comment by FrankC01** [comment]
Updated the GIST showing some progress however; GISTS is throwing a 500 (oops, perhaps it's my sloppy code)...

**Comment by vygr** [comment]
That gist is starting to look like real code. ;)

I have a sneaking suspicion that you have done programming before ;)

Gonna put an entry in my will that says you have to take over the project on my death. :)

**Comment by vygr** [comment]
(method Argument 'init (lambda (o v charg &optional htxt)
	(scall Argument 'init o v htxt)
	(def o
		'arg (or charg ())
		'type nil
		'default nil
		'acount 1)))

The (or charg ()) will give an error because () is going to throw “not a lambda”

Maybe ‘() or (list) ??

**Comment by vygr** [comment]
Also I’ve had a slaphead moment here thanks to seeing your use of:

(or arg default)

Duuuugh I am so stupid I added a function to do that ! (opt) in boot.inc :)

**Comment by vygr** [comment]
(method ChrysAP 'set_appname (lambda (_b app_name)
	"Application name setter"
	(set _b 'appname app_name)))

So your use here of ”Application name setter" is a comment ??

It’s going to be evaluated every time that function is called and then dropped on the floor.

If you really mean a comment use ;

Or have you got something in mind ?

**Comment by vygr** [comment]
(defun-bind parse-args (chap args)
	(if (find-str "-h" args)
		(parser-help chap)
		(print "Got Args" (slice 0 -1 args))))

(slice 0 -1 args) is just args?? If your meaning to take a copy of args, which I suspect you might be. You can use (cat args) to make a copy.

**Comment by vygr** [comment]
(defun-bind main ()
    ;initialize pipe details, command args and chrysarg procesor
    (defq stdio (create-stdio)
          args (stdio-get-args stdio)
          p (create-argparse "ChrysaLisp Argument Processor" "v0.0"))
    (parse-args p (slice 1 -1 args)))


Here you are not allowing for the possibility that the stdio from (create-stdio) could be nil due to some other member of the pipe failing !

I know this is some test code but just pointing this out. 

You will see most of the cmd/ apps use an and statement to break out if (create-stdio) returns nil.

**Comment by FrankC01** [comment]
Fixed:
```
(defun-bind main ()
    ;initialize pipe details, command args and chrysarg procesor
    (when (and
            (defq stdio (create-stdio))
            (defq args (stdio-get-args stdio))
            p (create-argument-processor))
        (parse-args p args)))
```

**Comment by FrankC01** [comment]
BTW, yes... I have coded in a good many languages and made reasonable progress on a new functional language. I won't say how long I've been at it but BAL was my first language....

**Comment by Schol-R-LEA** [comment]
> (method ChrysAP 'set_appname (lambda (_b app_name)
> "Application name setter"
> (set _b 'appname app_name)))
> 
> So your use here of ”Application name setter" is a comment ??
> 
> It’s going to be evaluated every time that function is called and then dropped on the floor.
> 
> If you really mean a comment use ;
> 
> Or have you got something in mind ?

While I am not the one who wrote that code, I think the intention was for it to be a docstring, which are widely used in Common Lisp, Clojure, and some Scheme dialects (as well as a few non-Lisp languages such as Python) and are intended for automatic code documentation generation (as opposed to code comments as such).  They are comparable to the Javadoc '/** ... */' comment blocks; the docgen tools scan the code for such instances, and pull them out to and use them for creating human-readable descriptions of the functions, usually either as UML diagrams or in a form such as:

_**Class ChrysAP**_
*The class docstring would go here* 

*Methods*
* *set_appname*
Application name setter
  **arguments**
  __b_ 
 _app_name_


The CL docgen tools take advantage of the fact that, as you said, a string in the function body but outside of a function application is simply ignored. Unlike comments, they are positional, and only valid as the first element of a defun - strings elsewhere in the function body are ignored by the docgen tools. 

From your comments, I take it that Chrysalisp doesn't use or support these.

**Comment by FrankC01** [comment]
@Schol-R-LEA I had a documentation outcome when I wrote the code, yes.

**Comment by vygr** [comment]
I was aware of its use but hadn't done so in ChrysaLisp. I have comments parsed out and auto create docs with them in a similar manner. The 'make doc' terminal app does this source scanning and doc building.

**Comment by FrankC01** [comment]
No worries, I'm focused on everything but documentation!

Anyway is `mcall` the public interface call and `scall` intended to be similar to `self` calls?

**Comment by vygr** [comment]
mcall is methods call, scall is super call. So call my parent class's method.

**Comment by vygr** [comment]
Generally user code don’t use scall but the class methods do use it.

Good example would be the use for ‘init and ‘deinit calling you to the parent class.

But this oops experiment from oops.lisp was an experiment. It’s not set in stone to do things like this !

**Comment by FrankC01** [comment]
Fully understand... making my through the macro capabilities of lisp or the subset you've implemented. I may be extending my copy of the class/method construct to add a wee bit of convenience and see how it works out.

**Comment by vygr** [comment]
Nuclearfall started an IRC channel a while back. I try to keep this open while I'm awake.

```
Nuclearfall has set up an IRC channel on irc.free node.net at #ChrysaLisp. The room
should be available 24-7 and I'll keep this open all the time I'm awake. Users
of Riot.im can also access the channel on matrix.org at #Chrysalisp:matrix.org

New Whiteboard app, with shamed face to Neauoire as it took so long to get
round to doing this. Despite it being my day job, or maybe because it's my day
job !
```

Chris

**Comment by FrankC01** [comment]
Made some progress on displaying help, building argument processing, etc. Next focus will be to actually do something on than display help

Updated [GIST](https://gist.github.com/FrankC01/fb30f3cc4baa5e8eb4a6a43c41d57e4dl)

**Comment by vygr** [comment]
I’m getting a 404 error on that link.

**Comment by FrankC01** [comment]
Try this [Link](https://gist.github.com/FrankC01/fb30f3cc4baa5e8eb4a6a43c41d57e4d)
